interop-testing/src/main/java/io/grpc/testing/integration/SameProcessPropagator.java

```diff
@@ -0,0 +1,121 @@
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.testing.integration;
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.Context;
+import io.grpc.Contexts;
+import io.grpc.ForwardingClientCall;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.ServerInterceptor;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Utility for propagating context objects between clients and servers when both are running
+ * in the same process. Suitable for testing only.
+ */
+public class SameProcessPropagator {
+
+  private static final AtomicInteger counter = new AtomicInteger();
+
+  @SuppressWarnings("unchecked")
+  private final Map<Context.Key, Metadata.Key<String>> keyToMetadataName;
+  private final Map<String, Object> recordedObjects = new HashMap<String, Object>();
+
+  /**
+   * Construct a propagator that will copy {@link Context} values from client to server for
+   * the given set of keys.
+   */
+  @SuppressWarnings("unchecked")
+  public SameProcessPropagator(Context.Key ...keys) {
+    keyToMetadataName = new HashMap<Context.Key, Metadata.Key<String>>();
+    for (Context.Key key : keys) {
+      keyToMetadataName.put(key,
+          Metadata.Key.of(key.toString() + "-" + counter.incrementAndGet(),
+              Metadata.ASCII_STRING_MARSHALLER));
+    }
+  }
+
+  /**
+   * Create a server interceptor that enables propagation on the server side.
+   */
+  public ServerInterceptor makeServerInterceptor() {
```

Comment:
I think I see what is trying to be gained by having `ServerContextTransformer`, but I'd prefer for users to continue implementing `ServerInterceptor` to remain orthogonal.

As an alternative, how would you feel about a method like:
```java
public class Contexts {
  public static <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
      Context context,
      MethodDescriptor<ReqT, RespT> method, ServerCall<RespT> call,
      Metadata headers, ServerCallHandler<ReqT, RespT> next) {
    Context old = context.attach();
    try {
      return new ContextualizedServerCallListener(next.startCall(method, call, headers), context);
    } finally {
      context.detach(old);
    }
  }
}
```

And then this method becomes:
```java
return new ServerInterceptor() {
  @Override
  public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
      MethodDescriptor<ReqT, RespT> method, ServerCall<RespT> call,
      Metadata headers, ServerCallHandler<ReqT, RespT> next) {
    Context context = Context.current();
    for (Map.Entry<Context.Key, Metadata.Key<String>> keyMeta : keyToMetadataName.entrySet()) {
      String objectId = headers.get(keyMeta.getValue());
      if (objectId != null) {
        context = context.withValue(keyMeta.getKey(), recordedObjects.get(objectId));
      }
    }
    return Contexts.interceptCall(context, method, call, headers, next);
  }
};
```